{\rtf1\ansi\ansicpg1252\cocoartf1504\cocoasubrtf830
{\fonttbl\f0\fnil\fcharset0 Monaco;}
{\colortbl;\red255\green255\blue255;\red242\green242\blue242;\red0\green0\blue0;\red46\green174\blue187;
\red47\green180\blue29;\red45\green150\blue30;\red47\green231\blue26;\red193\green101\blue28;\red180\green36\blue25;
\red64\green11\blue217;\red49\green139\blue238;\red161\green36\blue26;\red252\green33\blue24;}
{\*\expandedcolortbl;;\csgray\c95825;\csgray\c0\c85000;\cssrgb\c20198\c73241\c78251;
\cssrgb\c20239\c73899\c14948;\cssrgb\c20460\c63761\c15350;\cssrgb\c17864\c90514\c12812;\cssrgb\c80554\c47367\c13835;\cssrgb\c76410\c21699\c12524;
\cssrgb\c32309\c18668\c88228;\cssrgb\c23654\c62457\c94777;\cssrgb\c69952\c21067\c13205;\cssrgb\c100000\c23178\c11601;}
\margl1440\margr1440\vieww10800\viewh8400\viewkind0
\pard\tx560\tx1120\tx1680\tx2240\tx2800\tx3360\tx3920\tx4480\tx5040\tx5600\tx6160\tx6720\pardirnatural\partightenfactor0

\f0\fs26 \cf2 \cb3 \CocoaLigature0 python: can't open file '../manage.py': [Errno 2] No such file or directory\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py startapp book_authors\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ ls\
Dojo_ninjas_commands.rtf	db.sqlite3\
apps				dojo_ninjas\
book_authors			manage.py\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py makemigrations\
No changes detected\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py migrate\
\cf4 Operations to perform:\cf2 \
\cf1   Apply all migrations: \cf2 admin, auth, contenttypes, dojo_ninjas_app, sessions\
\cf4 Running migrations:\cf2 \
  No migrations to apply.\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py migrate\
\cf4 Operations to perform:\cf2 \
\cf1   Apply all migrations: \cf2 admin, auth, contenttypes, dojo_ninjas_app, sessions\
\cf4 Running migrations:\cf2 \
  No migrations to apply.\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ ls\
Dojo_ninjas_commands.rtf	db.sqlite3\
apps				dojo_ninjas\
book_authors			manage.py\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ cd apps\
(djangoEnv) Fifis-MBP:apps fifihome$ ls\
__init__.py	__init__.pyc	dojo_ninjas_app\
(djangoEnv) Fifis-MBP:apps fifihome$ ls\
__init__.py	__init__.pyc	dojo_ninjas_app\
(djangoEnv) Fifis-MBP:apps fifihome$ cd ../\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ ls\
Dojo_ninjas_commands.rtf	db.sqlite3\
apps				dojo_ninjas\
book_authors			manage.py\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ ls\
Dojo_ninjas_commands.rtf	db.sqlite3\
apps				dojo_ninjas\
book_authors			manage.py\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ cd apps\
(djangoEnv) Fifis-MBP:apps fifihome$ ls\
__init__.py	__init__.pyc	book_authors	dojo_ninjas_app\
(djangoEnv) Fifis-MBP:apps fifihome$ cd ../\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py makemigrations\
Traceback (most recent call last):\
  File "manage.py", line 22, in <module>\
    execute_from_command_line(sys.argv)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 363, in execute_from_command_line\
    utility.execute()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 337, in execute\
    django.setup()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/__init__.py", line 27, in setup\
    apps.populate(settings.INSTALLED_APPS)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/registry.py", line 85, in populate\
    app_config = AppConfig.create(entry)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/config.py", line 127, in create\
    import_module(entry)\
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module\
    __import__(name)\
ImportError: No module named books_authors\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py makemigrations\
Traceback (most recent call last):\
  File "manage.py", line 22, in <module>\
    execute_from_command_line(sys.argv)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 363, in execute_from_command_line\
    utility.execute()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 337, in execute\
    django.setup()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/__init__.py", line 27, in setup\
    apps.populate(settings.INSTALLED_APPS)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/registry.py", line 108, in populate\
    app_config.import_models()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/config.py", line 202, in import_models\
    self.models_module = import_module(models_module_name)\
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module\
    __import__(name)\
  File "/Users/fifihome/Desktop/Python/Django/dojo_ninjas/apps/book_authors/models.py", line 4, in <module>\
    class Book(models.Model):\
  File "/Users/fifihome/Desktop/Python/Django/dojo_ninjas/apps/book_authors/models.py", line 9, in Book\
    authors = models.ManytoManyField(Author, related_name = "authors")\
AttributeError: 'module' object has no attribute 'ManytoManyField'\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py makemigrations\
Traceback (most recent call last):\
  File "manage.py", line 22, in <module>\
    execute_from_command_line(sys.argv)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 363, in execute_from_command_line\
    utility.execute()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 337, in execute\
    django.setup()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/__init__.py", line 27, in setup\
    apps.populate(settings.INSTALLED_APPS)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/registry.py", line 108, in populate\
    app_config.import_models()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/config.py", line 202, in import_models\
    self.models_module = import_module(models_module_name)\
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module\
    __import__(name)\
  File "/Users/fifihome/Desktop/Python/Django/dojo_ninjas/apps/book_authors/models.py", line 4, in <module>\
    class Book(models.Model):\
  File "/Users/fifihome/Desktop/Python/Django/dojo_ninjas/apps/book_authors/models.py", line 9, in Book\
    authors = models.ManytoManyField(Author, related_name = "authors")\
AttributeError: 'module' object has no attribute 'ManytoManyField'\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py makemigrations\
Traceback (most recent call last):\
  File "manage.py", line 22, in <module>\
    execute_from_command_line(sys.argv)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 363, in execute_from_command_line\
    utility.execute()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 337, in execute\
    django.setup()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/__init__.py", line 27, in setup\
    apps.populate(settings.INSTALLED_APPS)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/registry.py", line 108, in populate\
    app_config.import_models()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/config.py", line 202, in import_models\
    self.models_module = import_module(models_module_name)\
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module\
    __import__(name)\
  File "/Users/fifihome/Desktop/Python/Django/dojo_ninjas/apps/book_authors/models.py", line 4, in <module>\
    class Book(models.Model):\
  File "/Users/fifihome/Desktop/Python/Django/dojo_ninjas/apps/book_authors/models.py", line 9, in Book\
    authors = models.ManytoManyField(Author, related_name = "authors")\
AttributeError: 'module' object has no attribute 'ManytoManyField'\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py makemigrations\
Traceback (most recent call last):\
  File "manage.py", line 22, in <module>\
    execute_from_command_line(sys.argv)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 363, in execute_from_command_line\
    utility.execute()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 337, in execute\
    django.setup()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/__init__.py", line 27, in setup\
    apps.populate(settings.INSTALLED_APPS)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/registry.py", line 108, in populate\
    app_config.import_models()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/config.py", line 202, in import_models\
    self.models_module = import_module(models_module_name)\
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module\
    __import__(name)\
  File "/Users/fifihome/Desktop/Python/Django/dojo_ninjas/apps/book_authors/models.py", line 12, in <module>\
    class Author(models.Model):\
  File "/Users/fifihome/Desktop/Python/Django/dojo_ninjas/apps/book_authors/models.py", line 18, in Author\
    book = models.ManytoManyField(Book, related_name = "books")\
AttributeError: 'module' object has no attribute 'ManytoManyField'\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py makemigrations\
Traceback (most recent call last):\
  File "manage.py", line 22, in <module>\
    execute_from_command_line(sys.argv)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 363, in execute_from_command_line\
    utility.execute()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 337, in execute\
    django.setup()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/__init__.py", line 27, in setup\
    apps.populate(settings.INSTALLED_APPS)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/registry.py", line 108, in populate\
    app_config.import_models()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/config.py", line 202, in import_models\
    self.models_module = import_module(models_module_name)\
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module\
    __import__(name)\
  File "/Users/fifihome/Desktop/Python/Django/dojo_ninjas/apps/book_authors/models.py", line 12, in <module>\
    class Author(models.Model):\
  File "/Users/fifihome/Desktop/Python/Django/dojo_ninjas/apps/book_authors/models.py", line 18, in Author\
    books = models.ManytoManyField(Book, related_name = "authors")\
AttributeError: 'module' object has no attribute 'ManytoManyField'\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py makemigrations\
Traceback (most recent call last):\
  File "manage.py", line 22, in <module>\
    execute_from_command_line(sys.argv)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 363, in execute_from_command_line\
    utility.execute()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 337, in execute\
    django.setup()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/__init__.py", line 27, in setup\
    apps.populate(settings.INSTALLED_APPS)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/registry.py", line 108, in populate\
    app_config.import_models()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/apps/config.py", line 202, in import_models\
    self.models_module = import_module(models_module_name)\
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module\
    __import__(name)\
  File "/Users/fifihome/Desktop/Python/Django/dojo_ninjas/apps/book_authors/models.py", line 12, in <module>\
    class Author(models.Model):\
  File "/Users/fifihome/Desktop/Python/Django/dojo_ninjas/apps/book_authors/models.py", line 16, in Author\
    books = models.ManytoManyField(Book, related_name = "authors")\
AttributeError: 'module' object has no attribute 'ManytoManyField'\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py makemigrations\
\cf4 Migrations for 'book_authors':\cf2 \
  \cf1 apps/book_authors/migrations/0001_initial.py\cf2 \
    - Create model Author\
    - Create model Book\
    - Add field books to author\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py migrate\
\cf4 Operations to perform:\cf2 \
\cf1   Apply all migrations: \cf2 admin, auth, book_authors, contenttypes, dojo_ninjas_app, sessions\
\cf4 Running migrations:\cf2 \
  Applying book_authors.0001_initial...\cf5  OK\cf2 \
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py shell\
\
^CTraceback (most recent call last):\
  File "manage.py", line 22, in <module>\
    execute_from_command_line(sys.argv)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 363, in execute_from_command_line\
    utility.execute()\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/__init__.py", line 355, in execute\
    self.fetch_command(subcommand).run_from_argv(self.argv)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/base.py", line 283, in run_from_argv\
    self.execute(*args, **cmd_options)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/base.py", line 330, in execute\
    output = self.handle(*args, **options)\
  File "/Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/core/management/commands/shell.py", line 101, in handle\
    exec(sys.stdin.read())\
KeyboardInterrupt\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ ls\
Dojo_ninjas_commands.rtf	dojo_ninjas\
apps				manage.py\
db.sqlite3\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py shell\
Python 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 12:39:47) \
Type "copyright", "credits" or "license" for more information.\
\
IPython 5.4.1 -- An enhanced Interactive Python.\
?         -> Introduction and overview of IPython's features.\
%quickref -> Quick reference.\
help      -> Python's own help system.\
object?   -> Details about 'object', use 'object??' for extra details.\
\
\cf6 In [\cf7 1\cf6 ]: \cf2 Book.objects.create(name=\cf8 "C Sharp"\cf2 )\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 NameError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-1-394e7a566208>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Book\cf10 .\cf2 objects\cf10 .\cf2 create\cf10 (\cf2 name\cf10 ="C Sharp")\cf2 \
\
\cf9 NameError\cf2 : name 'Book' is not defined\
\
\cf6 In [\cf7 2\cf6 ]: \cf2 Book.objects.create(name=\cf8 "C Sharp"\cf2 )\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 NameError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-2-394e7a566208>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Book\cf10 .\cf2 objects\cf10 .\cf2 create\cf10 (\cf2 name\cf10 ="C Sharp")\cf2 \
\
\cf9 NameError\cf2 : name 'Book' is not defined\
\
\cf6 In [\cf7 3\cf6 ]: \cf2 books.objects.create(name=\cf8 "C Sharp"\cf2 )\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 NameError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-3-6e7b7e9e5b62>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 books\cf10 .\cf2 objects\cf10 .\cf2 create\cf10 (\cf2 name\cf10 ="C Sharp")\cf2 \
\
\cf9 NameError\cf2 : name 'books' is not defined\
\
\cf6 In [\cf7 4\cf6 ]: \cf2 Book.objects.create(name=\cf8 "C Sharp"\cf2 )\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 NameError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-4-394e7a566208>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Book\cf10 .\cf2 objects\cf10 .\cf2 create\cf10 (\cf2 name\cf10 ="C Sharp")\cf2 \
\
\cf9 NameError\cf2 : name 'Book' is not defined\
\
\cf6 In [\cf7 5\cf6 ]: exit\cf2 ()\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ ls\
Dojo_ninjas_commands.rtf	dojo_ninjas\
apps				manage.py\
db.sqlite3\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ cd apps\
(djangoEnv) Fifis-MBP:apps fifihome$ ls\
__init__.py	__init__.pyc	book_authors	dojo_ninjas_app\
(djangoEnv) Fifis-MBP:apps fifihome$ cd book_authors\
(djangoEnv) Fifis-MBP:book_authors fifihome$ ls\
__init__.py	admin.py	apps.py		models.py	tests.py\
__init__.pyc	admin.pyc	migrations	models.pyc	views.py\
(djangoEnv) Fifis-MBP:book_authors fifihome$ cd ../\
(djangoEnv) Fifis-MBP:apps fifihome$ cd ../\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ ls\
Dojo_ninjas_commands.rtf	dojo_ninjas\
apps				manage.py\
db.sqlite3\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py shel\
Unknown command: 'shel'\
Type 'manage.py help' for usage.\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py shell\
Python 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 12:39:47) \
Type "copyright", "credits" or "license" for more information.\
\
IPython 5.4.1 -- An enhanced Interactive Python.\
?         -> Introduction and overview of IPython's features.\
%quickref -> Quick reference.\
help      -> Python's own help system.\
object?   -> Details about 'object', use 'object??' for extra details.\
\
\cf6 In [\cf7 1\cf6 ]: \cf2 Book.objects.create(name=\cf8 "C Sharp"\cf2 )\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 NameError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-1-394e7a566208>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Book\cf10 .\cf2 objects\cf10 .\cf2 create\cf10 (\cf2 name\cf10 ="C Sharp")\cf2 \
\
\cf9 NameError\cf2 : name 'Book' is not defined\
\
\cf6 In [\cf7 2\cf6 ]: import\cf2  \cf11 table\cf2  \cf11 data\cf2  \cf11 from\cf2  \cf11 apps.book_authors.models\cf2  \cf11 import\cf2  *\
\cf4   File \cf5 "<ipython-input-2-ab7cf4767c3b>"\cf4 , line \cf5 1\cf2 \
\cf9     import table data from apps.book_authors.models import *\cf2 \
                    ^\
\cf9 SyntaxError:\cf2  invalid syntax\
\
\
\cf6 In [\cf7 3\cf6 ]: from\cf2  \cf11 apps.book_authors.models\cf2  \cf6 import\cf2  *\
\
\cf6 In [\cf7 4\cf6 ]: \cf2 Book.objects.create(name=\cf8 "C Sharp"\cf2 )\
\cf12 Out[\cf13 4\cf12 ]: \cf2 <Book object: C Sharp SOME STRING>\
\
\cf6 In [\cf7 5\cf6 ]: \cf2 Book.objects.create(name=\cf8 "Java"\cf2 )\
\cf12 Out[\cf13 5\cf12 ]: \cf2 <Book object: Java SOME STRING>\
\
\cf6 In [\cf7 6\cf6 ]: \cf2 Book.objects.create(name=\cf8 "Python"\cf2 )\
\cf12 Out[\cf13 6\cf12 ]: \cf2 <Book object: Python SOME STRING>\
\
\cf6 In [\cf7 7\cf6 ]: \cf2 Book.objects.create(name=\cf8 "PHP"\cf2 )\
\cf12 Out[\cf13 7\cf12 ]: \cf2 <Book object: PHP SOME STRING>\
\
\cf6 In [\cf7 8\cf6 ]: \cf2 Book.objects.create(name=\cf8 "PHP"\cf2 )\
\
\cf12 Out[\cf13 8\cf12 ]: \cf2 <Book object: PHP SOME STRING>\
\
\cf6 In [\cf7 9\cf6 ]: \cf2 \
\
\cf6 In [\cf7 9\cf6 ]: \cf2 \
\
\cf6 In [\cf7 9\cf6 ]: \cf2 Book.objects.all()\
\cf12 Out[\cf13 9\cf12 ]: \cf2 <QuerySet [<Book object: C Sharp SOME STRING>, <Book object: Java SOME STRING>, <Book object: Python SOME STRING>, <Book object: PHP SOME STRING>, <Book object: PHP SOME STRING>]>\
\
\cf6 In [\cf7 10\cf6 ]: \cf2 Blog.objects.last().delete\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 NameError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-10-29cfb459b8e8>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Blog\cf10 .\cf2 objects\cf10 .\cf2 last\cf10 ().\cf2 delete\
\
\cf9 NameError\cf2 : name 'Blog' is not defined\
\
\cf6 In [\cf7 11\cf6 ]: \cf2 Book.objects.last().delete()\
\cf12 Out[\cf13 11\cf12 ]: \cf2 (1, \{u'book_authors.Author_books': 0, u'book_authors.Book': 1\})\
\
\cf6 In [\cf7 12\cf6 ]: \cf2 Book.objects.all()\
\cf12 Out[\cf13 12\cf12 ]: \cf2 <QuerySet [<Book object: C Sharp SOME STRING>, <Book object: Java SOME STRING>, <Book object: Python SOME STRING>, <Book object: PHP SOME STRING>]>\
\
\cf6 In [\cf7 13\cf6 ]: \cf2 Book.objects.create(name=\cf8 "Ruby"\cf2 )\
\cf12 Out[\cf13 13\cf12 ]: \cf2 <Book object: Ruby SOME STRING>\
\
\cf6 In [\cf7 14\cf6 ]: \cf2 Author.objects.create(name=\cf8 "Mike"\cf2 )\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 TypeError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-14-0608de0bb827>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Author\cf10 .\cf2 objects\cf10 .\cf2 create\cf10 (\cf2 name\cf10 ="Mike")\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/manager.pyc\cf2  in \cf4 manager_method\cf10 (self, *args, **kwargs)\cf2 \
\cf5      83\cf2          \cf5 def\cf2  create_method\cf10 (\cf2 name\cf10 ,\cf2  method\cf10 ):\cf2 \
\cf5      84\cf2              \cf5 def\cf2  manager_method\cf10 (\cf2 self\cf10 ,\cf2  \cf10 *\cf2 args\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 ):\cf2 \
\cf5 ---> 85\cf9                  \cf5 return\cf2  getattr\cf10 (\cf2 self\cf10 .\cf2 get_queryset\cf10 (),\cf2  name\cf10 )(*\cf2 args\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 )\cf2 \
\cf5      86\cf2              manager_method\cf10 .\cf2 __name__ \cf10 =\cf2  method\cf10 .\cf2 __name__\
\cf5      87\cf2              manager_method\cf10 .\cf2 __doc__ \cf10 =\cf2  method\cf10 .\cf2 __doc__\
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/query.pyc\cf2  in \cf4 create\cf10 (self, **kwargs)\cf2 \
\cf5     390\cf2          \cf5 and\cf2  returning the created object\cf10 .\cf2 \
\cf5     391\cf2          """\
\cf5 --> 392\cf9          \cf2 obj \cf10 =\cf2  self\cf10 .\cf2 model\cf10 (**\cf2 kwargs\cf10 )\cf2 \
\cf5     393\cf2          self\cf10 .\cf2 _for_write \cf10 =\cf2  True\
\cf5     394\cf2          obj\cf10 .\cf2 save\cf10 (\cf2 force_insert\cf10 =\cf2 True\cf10 ,\cf2  using\cf10 =\cf2 self\cf10 .\cf2 db\cf10 )\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/base.pyc\cf2  in \cf4 __init__\cf10 (self, *args, **kwargs)\cf2 \
\cf5     570\cf2                      \cf5 pass\cf2 \
\cf5     571\cf2              \cf5 if\cf2  kwargs\cf10 :\cf2 \
\cf5 --> 572\cf9                  \cf5 raise\cf2  TypeError\cf10 ("'%s' is an invalid keyword argument for this function"\cf2  \cf10 %\cf2  list\cf10 (\cf2 kwargs\cf10 )[\cf4 0\cf10 ])\cf2 \
\cf5     573\cf2          super\cf10 (\cf2 Model\cf10 ,\cf2  self\cf10 ).\cf2 __init__\cf10 ()\cf2 \
\cf5     574\cf2          post_init\cf10 .\cf2 send\cf10 (\cf2 sender\cf10 =\cf2 cls\cf10 ,\cf2  instance\cf10 =\cf2 self\cf10 )\cf2 \
\
\cf9 TypeError\cf2 : 'name' is an invalid keyword argument for this function\
\
\cf6 In [\cf7 15\cf6 ]: \cf2 Author.objects.create(first_name=\cf8 "Mike"\cf2 )\
\cf12 Out[\cf13 15\cf12 ]: \cf2 <Author object: Mike >\
\
\cf6 In [\cf7 16\cf6 ]: \cf2 Author.objects.create(first_name=\cf8 "Speros"\cf2 )\
\cf12 Out[\cf13 16\cf12 ]: \cf2 <Author object: Speros >\
\
\cf6 In [\cf7 17\cf6 ]: \cf2 Author.objects.create(first_name=\cf8 "John"\cf2 )\
\cf12 Out[\cf13 17\cf12 ]: \cf2 <Author object: John >\
\
\cf6 In [\cf7 18\cf6 ]: \cf2 Author.objects.create(first_name=\cf8 "Jadee"\cf2 )\
\cf12 Out[\cf13 18\cf12 ]: \cf2 <Author object: Jadee >\
\
\cf6 In [\cf7 19\cf6 ]: \cf2 Author.objects.create(first_name=\cf8 "Jay"\cf2 )\
\cf12 Out[\cf13 19\cf12 ]: \cf2 <Author object: Jay >\
\
\cf6 In [\cf7 20\cf6 ]: exit\cf2 ()\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ ls\
Dojo_ninjas_commands.rtf	dojo_ninjas\
apps				manage.py\
db.sqlite3\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py makemigrations\
\cf4 Migrations for 'book_authors':\cf2 \
  \cf1 apps/book_authors/migrations/0002_author_notes.py\cf2 \
    - Add field notes to author\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py migrate\
\cf4 Operations to perform:\cf2 \
\cf1   Apply all migrations: \cf2 admin, auth, book_authors, contenttypes, dojo_ninjas_app, sessions\
\cf4 Running migrations:\cf2 \
  Applying book_authors.0002_author_notes...\cf5  OK\cf2 \
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py shell\
Python 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 12:39:47) \
Type "copyright", "credits" or "license" for more information.\
\
IPython 5.4.1 -- An enhanced Interactive Python.\
?         -> Introduction and overview of IPython's features.\
%quickref -> Quick reference.\
help      -> Python's own help system.\
object?   -> Details about 'object', use 'object??' for extra details.\
\
\cf6 In [\cf7 1\cf6 ]: from\cf2  \cf11 apps.book_authors.models\cf2  \cf6 import\cf2  *\
\
\cf6 In [\cf7 2\cf6 ]: \cf2 Book.objects.all()\
\cf12 Out[\cf13 2\cf12 ]: \cf2 <QuerySet [<Book object: C Sharp SOME STRING>, <Book object: Java SOME STRING>, <Book object: Python SOME STRING>, <Book object: PHP SOME STRING>, <Book object: Ruby SOME STRING>]>\
\
\cf6 In [\cf7 3\cf6 ]: \cf2 b = Books.object.last()\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 NameError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-3-9fe323de7a2d>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 b \cf10 =\cf2  Books\cf10 .\cf2 object\cf10 .\cf2 last\cf10 ()\cf2 \
\
\cf9 NameError\cf2 : name 'Books' is not defined\
\
\cf6 In [\cf7 4\cf6 ]: \cf2 b = Book.objects.last()\
\
\cf6 In [\cf7 5\cf6 ]: \cf2 b.name = \cf8 "C#"\cf2 \
\
\cf6 In [\cf7 6\cf6 ]: \cf2 b.save()\
\
\cf6 In [\cf7 7\cf6 ]: \cf2 Book.objects.last()\
\cf12 Out[\cf13 7\cf12 ]: \cf2 <Book object: C# SOME STRING>\
\
\cf6 In [\cf7 8\cf6 ]: \cf2 Author.object.last()\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-8-39fb65d1c73c>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Author\cf10 .\cf2 object\cf10 .\cf2 last\cf10 ()\cf2 \
\
\cf9 AttributeError\cf2 : type object 'Author' has no attribute 'object'\
\
\cf6 In [\cf7 9\cf6 ]: \cf2 Author.objects.last()\
\cf12 Out[\cf13 9\cf12 ]: \cf2 <Author object: Jay >\
\
\cf6 In [\cf7 10\cf6 ]: \cf2 a = Author.objects.last()\
\
\cf6 In [\cf7 11\cf6 ]: \cf2 a.first_name = \cf8 "Ketul"\cf2 \
\
\cf6 In [\cf7 12\cf6 ]: \cf2 a.save()\
\
\cf6 In [\cf7 13\cf6 ]: \cf2 Author.objects.last()\
\cf12 Out[\cf13 13\cf12 ]: \cf2 <Author object: Ketul >\
\
\cf6 In [\cf7 14\cf6 ]: \cf2 Author.objects.create(book_id=Book.objects.get(\cf6 id\cf2 =\cf6 1\cf2 ), author = \cf8 "Mike"\cf2 )\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 TypeError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-14-7e58d35684bb>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Author\cf10 .\cf2 objects\cf10 .\cf2 create\cf10 (\cf2 book_id\cf10 =\cf2 Book\cf10 .\cf2 objects\cf10 .\cf2 get\cf10 (\cf2 id\cf10 =\cf4 1\cf10 ),\cf2  author \cf10 =\cf2  \cf10 "Mike")\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/manager.pyc\cf2  in \cf4 manager_method\cf10 (self, *args, **kwargs)\cf2 \
\cf5      83\cf2          \cf5 def\cf2  create_method\cf10 (\cf2 name\cf10 ,\cf2  method\cf10 ):\cf2 \
\cf5      84\cf2              \cf5 def\cf2  manager_method\cf10 (\cf2 self\cf10 ,\cf2  \cf10 *\cf2 args\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 ):\cf2 \
\cf5 ---> 85\cf9                  \cf5 return\cf2  getattr\cf10 (\cf2 self\cf10 .\cf2 get_queryset\cf10 (),\cf2  name\cf10 )(*\cf2 args\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 )\cf2 \
\cf5      86\cf2              manager_method\cf10 .\cf2 __name__ \cf10 =\cf2  method\cf10 .\cf2 __name__\
\cf5      87\cf2              manager_method\cf10 .\cf2 __doc__ \cf10 =\cf2  method\cf10 .\cf2 __doc__\
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/query.pyc\cf2  in \cf4 create\cf10 (self, **kwargs)\cf2 \
\cf5     390\cf2          \cf5 and\cf2  returning the created object\cf10 .\cf2 \
\cf5     391\cf2          """\
\cf5 --> 392\cf9          \cf2 obj \cf10 =\cf2  self\cf10 .\cf2 model\cf10 (**\cf2 kwargs\cf10 )\cf2 \
\cf5     393\cf2          self\cf10 .\cf2 _for_write \cf10 =\cf2  True\
\cf5     394\cf2          obj\cf10 .\cf2 save\cf10 (\cf2 force_insert\cf10 =\cf2 True\cf10 ,\cf2  using\cf10 =\cf2 self\cf10 .\cf2 db\cf10 )\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/base.pyc\cf2  in \cf4 __init__\cf10 (self, *args, **kwargs)\cf2 \
\cf5     570\cf2                      \cf5 pass\cf2 \
\cf5     571\cf2              \cf5 if\cf2  kwargs\cf10 :\cf2 \
\cf5 --> 572\cf9                  \cf5 raise\cf2  TypeError\cf10 ("'%s' is an invalid keyword argument for this function"\cf2  \cf10 %\cf2  list\cf10 (\cf2 kwargs\cf10 )[\cf4 0\cf10 ])\cf2 \
\cf5     573\cf2          super\cf10 (\cf2 Model\cf10 ,\cf2  self\cf10 ).\cf2 __init__\cf10 ()\cf2 \
\cf5     574\cf2          post_init\cf10 .\cf2 send\cf10 (\cf2 sender\cf10 =\cf2 cls\cf10 ,\cf2  instance\cf10 =\cf2 self\cf10 )\cf2 \
\
\cf9 TypeError\cf2 : 'book_id' is an invalid keyword argument for this function\
\
\cf6 In [\cf7 15\cf6 ]: \cf2 Author.objects.create(books=Book.objects.get(\cf6 id\cf2 =\cf6 1\cf2 ), author = \cf8 "Mike"\cf2 )\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 ValueError\cf2                                 Traceback (most recent call last)\
\cf5 <ipython-input-15-371cc15dfa05>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Author\cf10 .\cf2 objects\cf10 .\cf2 create\cf10 (\cf2 books\cf10 =\cf2 Book\cf10 .\cf2 objects\cf10 .\cf2 get\cf10 (\cf2 id\cf10 =\cf4 1\cf10 ),\cf2  author \cf10 =\cf2  \cf10 "Mike")\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/manager.pyc\cf2  in \cf4 manager_method\cf10 (self, *args, **kwargs)\cf2 \
\cf5      83\cf2          \cf5 def\cf2  create_method\cf10 (\cf2 name\cf10 ,\cf2  method\cf10 ):\cf2 \
\cf5      84\cf2              \cf5 def\cf2  manager_method\cf10 (\cf2 self\cf10 ,\cf2  \cf10 *\cf2 args\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 ):\cf2 \
\cf5 ---> 85\cf9                  \cf5 return\cf2  getattr\cf10 (\cf2 self\cf10 .\cf2 get_queryset\cf10 (),\cf2  name\cf10 )(*\cf2 args\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 )\cf2 \
\cf5      86\cf2              manager_method\cf10 .\cf2 __name__ \cf10 =\cf2  method\cf10 .\cf2 __name__\
\cf5      87\cf2              manager_method\cf10 .\cf2 __doc__ \cf10 =\cf2  method\cf10 .\cf2 __doc__\
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/query.pyc\cf2  in \cf4 create\cf10 (self, **kwargs)\cf2 \
\cf5     390\cf2          \cf5 and\cf2  returning the created object\cf10 .\cf2 \
\cf5     391\cf2          """\
\cf5 --> 392\cf9          \cf2 obj \cf10 =\cf2  self\cf10 .\cf2 model\cf10 (**\cf2 kwargs\cf10 )\cf2 \
\cf5     393\cf2          self\cf10 .\cf2 _for_write \cf10 =\cf2  True\
\cf5     394\cf2          obj\cf10 .\cf2 save\cf10 (\cf2 force_insert\cf10 =\cf2 True\cf10 ,\cf2  using\cf10 =\cf2 self\cf10 .\cf2 db\cf10 )\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/base.pyc\cf2  in \cf4 __init__\cf10 (self, *args, **kwargs)\cf2 \
\cf5     565\cf2                      \cf5 if\cf2  prop \cf5 in\cf2  property_names \cf5 or\cf2  opts\cf10 .\cf2 get_field\cf10 (\cf2 prop\cf10 ):\cf2 \
\cf5     566\cf2                          \cf5 if\cf2  kwargs\cf10 [\cf2 prop\cf10 ]\cf2  \cf5 is\cf2  \cf5 not\cf2  _DEFERRED\cf10 :\cf2 \
\cf5 --> 567\cf9                              \cf2 _setattr\cf10 (\cf2 self\cf10 ,\cf2  prop\cf10 ,\cf2  kwargs\cf10 [\cf2 prop\cf10 ])\cf2 \
\cf5     568\cf2                          \cf5 del\cf2  kwargs\cf10 [\cf2 prop\cf10 ]\cf2 \
\cf5     569\cf2                  \cf5 except\cf2  \cf10 (\cf2 AttributeError\cf10 ,\cf2  FieldDoesNotExist\cf10 ):\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.pyc\cf2  in \cf4 __set__\cf10 (self, instance, value)\cf2 \
\cf5     534\cf2              RemovedInDjango20Warning\cf10 ,\cf2  stacklevel\cf10 =\cf4 2\cf10 ,\cf2 \
\cf5     535\cf2          )\
\cf5 --> 536\cf9          \cf2 manager \cf10 =\cf2  self\cf10 .\cf2 __get__\cf10 (\cf2 instance\cf10 )\cf2 \
\cf5     537\cf2          manager\cf10 .\cf2 set\cf10 (\cf2 value\cf10 )\cf2 \
\cf5     538\cf2  \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.pyc\cf2  in \cf4 __get__\cf10 (self, instance, cls)\cf2 \
\cf5     511\cf2              \cf5 return\cf2  self\
\cf5     512\cf2  \
\cf5 --> 513\cf9          \cf5 return\cf2  self\cf10 .\cf2 related_manager_cls\cf10 (\cf2 instance\cf10 )\cf2 \
\cf5     514\cf2  \
\cf5     515\cf2      \cf5 def\cf2  _get_set_deprecation_msg_params\cf10 (\cf2 self\cf10 ):\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.pyc\cf2  in \cf4 __init__\cf10 (self, instance)\cf2 \
\cf5     828\cf2                  raise ValueError('"%r" needs to have a value for field "%s" before '\
\cf5     829\cf2                                   \cf10 'this many-to-many relationship can be used.'\cf2  \cf10 %\cf2 \
\cf5 --> 830\cf9                                   (instance, self.pk_field_names[self.source_field_name]))\cf2 \
\cf5     831\cf2              \cf9 # Even if this relation is not to pk, we require still pk value.\cf2 \
\cf5     832\cf2              \cf9 # The wish is that the instance has been already saved to DB,\cf2 \
\
\cf9 ValueError\cf2 : "<Author object:  >" needs to have a value for field "id" before this many-to-many relationship can be used.\
\
\cf6 In [\cf7 16\cf6 ]: \cf2 Author.objects.create(books=Book.objects.get(\cf6 id\cf2 =\cf6 1\cf2 ), Author.object.get(\cf6 id\cf2 =\cf6 1\cf2 ))\
\cf4   File \cf5 "<ipython-input-16-936ec2c098dd>"\cf4 , line \cf5 1\cf2 \
\cf9     Author.objects.create(books=Book.objects.get(id=1), Author.object.get(id=1))\cf2 \
\cf9 SyntaxError:\cf2  non-keyword arg after keyword arg\
\
\
\cf6 In [\cf7 17\cf6 ]: \cf2 Author.objects.create(books=Book.objects.get(\cf6 id\cf2 =\cf6 1\cf2 ), author = Author.object.get(\cf6 i\cf2 \
\cf6     ...: d\cf2 =\cf6 1\cf2 ))\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-17-e1ebca631fb1>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Author\cf10 .\cf2 objects\cf10 .\cf2 create\cf10 (\cf2 books\cf10 =\cf2 Book\cf10 .\cf2 objects\cf10 .\cf2 get\cf10 (\cf2 id\cf10 =\cf4 1\cf10 ),\cf2  author \cf10 =\cf2  Author\cf10 .\cf2 object\cf10 .\cf2 get\cf10 (\cf2 id\cf10 =\cf4 1\cf10 ))\cf2 \
\
\cf9 AttributeError\cf2 : type object 'Author' has no attribute 'object'\
\
\cf6 In [\cf7 18\cf6 ]: \cf2 Author.objects.create(books=Book.objects.get(\cf6 id\cf2 =\cf6 1\cf2 ), author = Author.objects.get(\
\cf6     ...: id\cf2 =\cf6 1\cf2 ))\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 ValueError\cf2                                 Traceback (most recent call last)\
\cf5 <ipython-input-18-b71f5daa8910>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Author\cf10 .\cf2 objects\cf10 .\cf2 create\cf10 (\cf2 books\cf10 =\cf2 Book\cf10 .\cf2 objects\cf10 .\cf2 get\cf10 (\cf2 id\cf10 =\cf4 1\cf10 ),\cf2  author \cf10 =\cf2  Author\cf10 .\cf2 objects\cf10 .\cf2 get\cf10 (\cf2 id\cf10 =\cf4 1\cf10 ))\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/manager.pyc\cf2  in \cf4 manager_method\cf10 (self, *args, **kwargs)\cf2 \
\cf5      83\cf2          \cf5 def\cf2  create_method\cf10 (\cf2 name\cf10 ,\cf2  method\cf10 ):\cf2 \
\cf5      84\cf2              \cf5 def\cf2  manager_method\cf10 (\cf2 self\cf10 ,\cf2  \cf10 *\cf2 args\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 ):\cf2 \
\cf5 ---> 85\cf9                  \cf5 return\cf2  getattr\cf10 (\cf2 self\cf10 .\cf2 get_queryset\cf10 (),\cf2  name\cf10 )(*\cf2 args\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 )\cf2 \
\cf5      86\cf2              manager_method\cf10 .\cf2 __name__ \cf10 =\cf2  method\cf10 .\cf2 __name__\
\cf5      87\cf2              manager_method\cf10 .\cf2 __doc__ \cf10 =\cf2  method\cf10 .\cf2 __doc__\
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/query.pyc\cf2  in \cf4 create\cf10 (self, **kwargs)\cf2 \
\cf5     390\cf2          \cf5 and\cf2  returning the created object\cf10 .\cf2 \
\cf5     391\cf2          """\
\cf5 --> 392\cf9          \cf2 obj \cf10 =\cf2  self\cf10 .\cf2 model\cf10 (**\cf2 kwargs\cf10 )\cf2 \
\cf5     393\cf2          self\cf10 .\cf2 _for_write \cf10 =\cf2  True\
\cf5     394\cf2          obj\cf10 .\cf2 save\cf10 (\cf2 force_insert\cf10 =\cf2 True\cf10 ,\cf2  using\cf10 =\cf2 self\cf10 .\cf2 db\cf10 )\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/base.pyc\cf2  in \cf4 __init__\cf10 (self, *args, **kwargs)\cf2 \
\cf5     565\cf2                      \cf5 if\cf2  prop \cf5 in\cf2  property_names \cf5 or\cf2  opts\cf10 .\cf2 get_field\cf10 (\cf2 prop\cf10 ):\cf2 \
\cf5     566\cf2                          \cf5 if\cf2  kwargs\cf10 [\cf2 prop\cf10 ]\cf2  \cf5 is\cf2  \cf5 not\cf2  _DEFERRED\cf10 :\cf2 \
\cf5 --> 567\cf9                              \cf2 _setattr\cf10 (\cf2 self\cf10 ,\cf2  prop\cf10 ,\cf2  kwargs\cf10 [\cf2 prop\cf10 ])\cf2 \
\cf5     568\cf2                          \cf5 del\cf2  kwargs\cf10 [\cf2 prop\cf10 ]\cf2 \
\cf5     569\cf2                  \cf5 except\cf2  \cf10 (\cf2 AttributeError\cf10 ,\cf2  FieldDoesNotExist\cf10 ):\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.pyc\cf2  in \cf4 __set__\cf10 (self, instance, value)\cf2 \
\cf5     534\cf2              RemovedInDjango20Warning\cf10 ,\cf2  stacklevel\cf10 =\cf4 2\cf10 ,\cf2 \
\cf5     535\cf2          )\
\cf5 --> 536\cf9          \cf2 manager \cf10 =\cf2  self\cf10 .\cf2 __get__\cf10 (\cf2 instance\cf10 )\cf2 \
\cf5     537\cf2          manager\cf10 .\cf2 set\cf10 (\cf2 value\cf10 )\cf2 \
\cf5     538\cf2  \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.pyc\cf2  in \cf4 __get__\cf10 (self, instance, cls)\cf2 \
\cf5     511\cf2              \cf5 return\cf2  self\
\cf5     512\cf2  \
\cf5 --> 513\cf9          \cf5 return\cf2  self\cf10 .\cf2 related_manager_cls\cf10 (\cf2 instance\cf10 )\cf2 \
\cf5     514\cf2  \
\cf5     515\cf2      \cf5 def\cf2  _get_set_deprecation_msg_params\cf10 (\cf2 self\cf10 ):\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.pyc\cf2  in \cf4 __init__\cf10 (self, instance)\cf2 \
\cf5     828\cf2                  raise ValueError('"%r" needs to have a value for field "%s" before '\
\cf5     829\cf2                                   \cf10 'this many-to-many relationship can be used.'\cf2  \cf10 %\cf2 \
\cf5 --> 830\cf9                                   (instance, self.pk_field_names[self.source_field_name]))\cf2 \
\cf5     831\cf2              \cf9 # Even if this relation is not to pk, we require still pk value.\cf2 \
\cf5     832\cf2              \cf9 # The wish is that the instance has been already saved to DB,\cf2 \
\
\cf9 ValueError\cf2 : "<Author object:  >" needs to have a value for field "id" before this many-to-many relationship can be used.\
\
\cf6 In [\cf7 19\cf6 ]: exit\cf2 ()\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ ls\
Dojo_ninjas_commands.rtf	dojo_ninjas\
apps				manage.py\
db.sqlite3\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py makemigrations\
\cf4 Migrations for 'book_authors':\cf2 \
  \cf1 apps/book_authors/migrations/0003_auto_20170817_1501.py\cf2 \
    - Alter field books on author\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py migrate\
\cf4 Operations to perform:\cf2 \
\cf1   Apply all migrations: \cf2 admin, auth, book_authors, contenttypes, dojo_ninjas_app, sessions\
\cf4 Running migrations:\cf2 \
  Applying book_authors.0003_auto_20170817_1501...\cf5  OK\cf2 \
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py shell\
Python 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 12:39:47) \
Type "copyright", "credits" or "license" for more information.\
\
IPython 5.4.1 -- An enhanced Interactive Python.\
?         -> Introduction and overview of IPython's features.\
%quickref -> Quick reference.\
help      -> Python's own help system.\
object?   -> Details about 'object', use 'object??' for extra details.\
\
\cf6 In [\cf7 1\cf6 ]: from\cf2  \cf11 apps.book_authors.models\cf2  \cf6 import\cf2  *\
\
\cf6 In [\cf7 2\cf6 ]: \cf2 Book.objects.first()\
\cf12 Out[\cf13 2\cf12 ]: \cf2 <Book object: C Sharp SOME STRING>\
\
\cf6 In [\cf7 3\cf6 ]: \cf2 a = Book.objects.first()\
\
\cf6 In [\cf7 4\cf6 ]: \cf2 a.save()\
\
\cf6 In [\cf7 5\cf6 ]: \cf2 a1 = Book.objects.get(\cf6 id\cf2 =\cf6 2\cf2 )\
\
\cf6 In [\cf7 6\cf6 ]: \cf2 a1.save\
\cf12 Out[\cf13 6\cf12 ]: \cf2 <bound method Book.save of <Book object: Java SOME STRING>>\
\
\cf6 In [\cf7 7\cf6 ]: \cf2 a\
\cf12 Out[\cf13 7\cf12 ]: \cf2 <Book object: C Sharp SOME STRING>\
\
\cf6 In [\cf7 8\cf6 ]: \cf2 a1\
\cf12 Out[\cf13 8\cf12 ]: \cf2 <Book object: Java SOME STRING>\
\
\cf6 In [\cf7 9\cf6 ]: \cf2 b = Author.objects(first)\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 NameError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-9-2b4c8179d30d>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 b \cf10 =\cf2  Author\cf10 .\cf2 objects\cf10 (\cf2 first\cf10 )\cf2 \
\
\cf9 NameError\cf2 : name 'first' is not defined\
\
\cf6 In [\cf7 10\cf6 ]: \cf2 b = Author.objects.first()\
\
\cf6 In [\cf7 11\cf6 ]: \cf2 b.save\
\cf12 Out[\cf13 11\cf12 ]: \cf2 <bound method Author.save of <Author object: Mike >>\
\
\cf6 In [\cf7 12\cf6 ]: \cf2 a.Book.add(b)\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-12-f7d92c12b4c7>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 a\cf10 .\cf2 Book\cf10 .\cf2 add\cf10 (\cf2 b\cf10 )\cf2 \
\
\cf9 AttributeError\cf2 : 'Book' object has no attribute 'Book'\
\
\cf6 In [\cf7 13\cf6 ]: \cf2 a.books.add(b)\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-13-610dbaae3336>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 a\cf10 .\cf2 books\cf10 .\cf2 add\cf10 (\cf2 b\cf10 )\cf2 \
\
\cf9 AttributeError\cf2 : 'Book' object has no attribute 'books'\
\
\cf6 In [\cf7 14\cf6 ]: \cf2 b.books.add(a)\
\
\cf6 In [\cf7 15\cf6 ]: \cf2 b.books.add(a1)\
\
\cf6 In [\cf7 16\cf6 ]: \cf2 b.books.save()\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-16-53f452a7d889>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 b\cf10 .\cf2 books\cf10 .\cf2 save\cf10 ()\cf2 \
\
\cf9 AttributeError\cf2 : 'ManyRelatedManager' object has no attribute 'save'\
\
\cf6 In [\cf7 17\cf6 ]: \cf2 books.save()\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 NameError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-17-a0e58ce306d9>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 books\cf10 .\cf2 save\cf10 ()\cf2 \
\
\cf9 NameError\cf2 : name 'books' is not defined\
\
\cf6 In [\cf7 18\cf6 ]: \cf2 b.save()\
\
\cf6 In [\cf7 19\cf6 ]: \cf2 Book.objects.all()\
\cf12 Out[\cf13 19\cf12 ]: \cf2 <QuerySet [<Book object: C Sharp SOME STRING>, <Book object: Java SOME STRING>, <Book object: Python SOME STRING>, <Book object: PHP SOME STRING>, <Book object: C# SOME STRING>]>\
\
\cf6 In [\cf7 20\cf6 ]: \cf2 Book.objects.first()\
\cf12 Out[\cf13 20\cf12 ]: \cf2 <Book object: C Sharp SOME STRING>\
\
\cf6 In [\cf7 21\cf6 ]: exit\cf2 ()\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py makemigrations\
\cf4 Migrations for 'book_authors':\cf2 \
  \cf1 apps/book_authors/migrations/0004_auto_20170817_1516.py\cf2 \
    - Alter field books on author\
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py migrate\
\cf4 Operations to perform:\cf2 \
\cf1   Apply all migrations: \cf2 admin, auth, book_authors, contenttypes, dojo_ninjas_app, sessions\
\cf4 Running migrations:\cf2 \
  Applying book_authors.0004_auto_20170817_1516...\cf5  OK\cf2 \
(djangoEnv) Fifis-MBP:dojo_ninjas fifihome$ python manage.py shell\
Python 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 12:39:47) \
Type "copyright", "credits" or "license" for more information.\
\
IPython 5.4.1 -- An enhanced Interactive Python.\
?         -> Introduction and overview of IPython's features.\
%quickref -> Quick reference.\
help      -> Python's own help system.\
object?   -> Details about 'object', use 'object??' for extra details.\
\
\cf6 In [\cf7 1\cf6 ]: from\cf2  \cf11 apps.book_authors.models\cf2  \cf6 import\cf2  *\
\
\cf6 In [\cf7 2\cf6 ]: \cf2 Book.objects.first().books_authors.all()\
\cf12 Out[\cf13 2\cf12 ]: \cf2 <QuerySet [<Author object: Mike >]>\
\
\cf6 In [\cf7 3\cf6 ]: \cf2 Author.objects.first().books_authors.all()\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-3-81efd05e2cca>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Author\cf10 .\cf2 objects\cf10 .\cf2 first\cf10 ().\cf2 books_authors\cf10 .\cf2 all\cf10 ()\cf2 \
\
\cf9 AttributeError\cf2 : 'Author' object has no attribute 'books_authors'\
\
\cf6 In [\cf7 4\cf6 ]: \cf2 Author.objects.first().books.all()\
\cf12 Out[\cf13 4\cf12 ]: \cf2 <QuerySet [<Book object: C Sharp SOME STRING>, <Book object: Java SOME STRING>]>\
\
\cf6 In [\cf7 5\cf6 ]: \cf2 b = Author.object.get(\cf6 id\cf2 =\cf6 2\cf2 )\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-5-08ff4c825099>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 b \cf10 =\cf2  Author\cf10 .\cf2 object\cf10 .\cf2 get\cf10 (\cf2 id\cf10 =\cf4 2\cf10 )\cf2 \
\
\cf9 AttributeError\cf2 : type object 'Author' has no attribute 'object'\
\
\cf6 In [\cf7 6\cf6 ]: \cf2 b = Author.objects.get(\cf6 id\cf2 =\cf6 2\cf2 )\
\
\cf6 In [\cf7 7\cf6 ]: \cf2 b.all()\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-7-5291ec0d1afb>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 b\cf10 .\cf2 all\cf10 ()\cf2 \
\
\cf9 AttributeError\cf2 : 'Author' object has no attribute 'all'\
\
\cf6 In [\cf7 8\cf6 ]: \cf2 b2 = Author.objects.get(\cf6 id\cf2 =\cf6 2\cf2 )\
\
\cf6 In [\cf7 9\cf6 ]: \cf2 Author.objects.get(\cf6 id\cf2 =\cf6 2\cf2 )\
\cf12 Out[\cf13 9\cf12 ]: \cf2 <Author object: Speros >\
\
\cf6 In [\cf7 10\cf6 ]: \cf2 b2.Book.add(a)\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-10-1212a79bb06f>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 b2\cf10 .\cf2 Book\cf10 .\cf2 add\cf10 (\cf2 a\cf10 )\cf2 \
\
\cf9 AttributeError\cf2 : 'Author' object has no attribute 'Book'\
\
\cf6 In [\cf7 11\cf6 ]: \cf2 b2.save()\
\
\cf6 In [\cf7 12\cf6 ]: \cf2 b2.books.add(a)\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 NameError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-12-6e5015dd0eea>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 b2\cf10 .\cf2 books\cf10 .\cf2 add\cf10 (\cf2 a\cf10 )\cf2 \
\
\cf9 NameError\cf2 : name 'a' is not defined\
\
\cf6 In [\cf7 13\cf6 ]: \cf2 a = Book.object.get(\cf6 id\cf2 =\cf6 1\cf2 )\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-13-367d7070cb44>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 a \cf10 =\cf2  Book\cf10 .\cf2 object\cf10 .\cf2 get\cf10 (\cf2 id\cf10 =\cf4 1\cf10 )\cf2 \
\
\cf9 AttributeError\cf2 : type object 'Book' has no attribute 'object'\
\
\cf6 In [\cf7 14\cf6 ]: \cf2 a = Book.objects.get(\cf6 id\cf2 =\cf6 1\cf2 )\
\
\cf6 In [\cf7 15\cf6 ]: \cf2 a.save()\
\
\cf6 In [\cf7 16\cf6 ]: \cf2 b2.books.add(a)\
\
\cf6 In [\cf7 17\cf6 ]: \cf2 a2 = Book.objects.get(\cf6 id\cf2 =\cf6 2\cf2 )\
\
\cf6 In [\cf7 18\cf6 ]: \cf2 a2.save()\
\
\cf6 In [\cf7 19\cf6 ]: \cf2 b2.books.add(a2)\
\
\cf6 In [\cf7 20\cf6 ]: \cf2 b2.save()\
\
\cf6 In [\cf7 21\cf6 ]: \cf2 a3 = Book.objects.get(\cf6 id\cf2 =\cf6 3\cf2 )\
\
\cf6 In [\cf7 22\cf6 ]: \cf2 a3.save()\
\
\cf6 In [\cf7 23\cf6 ]: \cf2 b2.books.add(a3)\
\
\cf6 In [\cf7 24\cf6 ]: \cf2 b2.save()\
\
\cf6 In [\cf7 25\cf6 ]: \cf2 Book.objects.first().books_authors.all()\
\cf12 Out[\cf13 25\cf12 ]: \cf2 <QuerySet [<Author object: Mike >, <Author object: Speros >]>\
\
\cf6 In [\cf7 26\cf6 ]: \cf2 Book.objects.get(\cf6 id\cf2 =\cf6 2\cf2 ).books_authors.all()\
\cf12 Out[\cf13 26\cf12 ]: \cf2 <QuerySet [<Author object: Mike >, <Author object: Speros >]>\
\
\cf6 In [\cf7 27\cf6 ]: \cf2 Book.objects.get(\cf6 id\cf2 =\cf6 3\cf2 ).books_authors.all()\
\cf12 Out[\cf13 27\cf12 ]: \cf2 <QuerySet [<Author object: Speros >]>\
\
\cf6 In [\cf7 28\cf6 ]: \cf2 a = Book.objects.get(\cf6 id\cf2 =\cf6 1\cf2 )\
\
\cf6 In [\cf7 29\cf6 ]: \cf2 a.save\
\cf12 Out[\cf13 29\cf12 ]: \cf2 <bound method Book.save of <Book object: C Sharp SOME STRING>>\
\
\cf6 In [\cf7 30\cf6 ]: \cf2 a.save()\
\
\cf6 In [\cf7 31\cf6 ]: \cf2 b3 = Author.objects.get(\cf6 id\cf2 =\cf6 3\cf2 )\
\
\cf6 In [\cf7 32\cf6 ]: \cf2 b3.save\
\cf12 Out[\cf13 32\cf12 ]: \cf2 <bound method Author.save of <Author object: John >>\
\
\cf6 In [\cf7 33\cf6 ]: \cf2 b3.save()\
\
\cf6 In [\cf7 34\cf6 ]: \cf2 b3.books.add(a)\
\
\cf6 In [\cf7 35\cf6 ]: \cf2 b3.save()\
\
\cf6 In [\cf7 36\cf6 ]: \cf2 a2 = Book.objects.get(\cf6 id\cf2 =\cf6 2\cf2 )\
\
\cf6 In [\cf7 37\cf6 ]: \cf2 a2.save()\
\
\cf6 In [\cf7 38\cf6 ]: \cf2 b3.books.add(a2)\
\
\cf6 In [\cf7 39\cf6 ]: \cf2 b3.save()\
\
\cf6 In [\cf7 40\cf6 ]: \cf2 a3 = Book.objects.get(\cf6 id\cf2 =\cf6 3\cf2 )\
\
\cf6 In [\cf7 41\cf6 ]: \cf2 a3.save()\
\
\cf6 In [\cf7 42\cf6 ]: \cf2 b3.books.add(a3)\
\
\cf6 In [\cf7 43\cf6 ]: \cf2 b3.save()\
\
\cf6 In [\cf7 44\cf6 ]: \cf2 a4 = Book.objects.get(\cf6 id\cf2 =\cf6 4\cf2 )\
\
\cf6 In [\cf7 45\cf6 ]: \cf2 a4.save\
\cf12 Out[\cf13 45\cf12 ]: \cf2 <bound method Book.save of <Book object: PHP SOME STRING>>\
\
\cf6 In [\cf7 46\cf6 ]: \cf2 a4.save()\
\
\cf6 In [\cf7 47\cf6 ]: \cf2 b3.books.add(a4)\
\
\cf6 In [\cf7 48\cf6 ]: \cf2 b3.save()-\
\cf4   File \cf5 "<ipython-input-48-3dbdda1589e4>"\cf4 , line \cf5 1\cf2 \
\cf9     b3.save()-\cf2 \
              ^\
\cf9 SyntaxError:\cf2  invalid syntax\
\
\
\cf6 In [\cf7 49\cf6 ]: \cf2 b3.save()\
\
\cf6 In [\cf7 50\cf6 ]: \cf2 b4 = Author.objects.last()\
\
\cf6 In [\cf7 51\cf6 ]: \cf2 b4.save()\
\
\cf6 In [\cf7 52\cf6 ]: \cf2 b4.books.add(Book.objects.all())\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 TypeError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-52-b7e5ca8991e2>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 b4\cf10 .\cf2 books\cf10 .\cf2 add\cf10 (\cf2 Book\cf10 .\cf2 objects\cf10 .\cf2 all\cf10 ())\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.pyc\cf2  in \cf4 add\cf10 (self, *objs)\cf2 \
\cf5     932\cf2              db \cf10 =\cf2  router\cf10 .\cf2 db_for_write\cf10 (\cf2 self\cf10 .\cf2 through\cf10 ,\cf2  instance\cf10 =\cf2 self\cf10 .\cf2 instance\cf10 )\cf2 \
\cf5     933\cf2              \cf5 with\cf2  transaction\cf10 .\cf2 atomic\cf10 (\cf2 using\cf10 =\cf2 db\cf10 ,\cf2  savepoint\cf10 =\cf2 False\cf10 ):\cf2 \
\cf5 --> 934\cf9                  \cf2 self\cf10 .\cf2 _add_items\cf10 (\cf2 self\cf10 .\cf2 source_field_name\cf10 ,\cf2  self\cf10 .\cf2 target_field_name\cf10 ,\cf2  \cf10 *\cf2 objs\cf10 )\cf2 \
\cf5     935\cf2  \
\cf5     936\cf2                  \cf9 # If this is a symmetrical m2m relation to self, add the mirror entry in the m2m table\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.pyc\cf2  in \cf4 _add_items\cf10 (self, source_field_name, target_field_name, *objs)\cf2 \
\cf5    1081\cf2                          .filter(**\{\
\cf5    1082\cf2                              source_field_name\cf10 :\cf2  self\cf10 .\cf2 related_val\cf10 [\cf4 0\cf10 ],\cf2 \
\cf5 -> 1083\cf9                              \cf10 '%s__in'\cf2  \cf10 %\cf2  target_field_name\cf10 :\cf2  new_ids\cf10 ,\cf2 \
\cf5    1084\cf2                          \}))\
\cf5    1085\cf2                  new_ids \cf10 =\cf2  new_ids \cf10 -\cf2  set\cf10 (\cf2 vals\cf10 )\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/query.pyc\cf2  in \cf4 filter\cf10 (self, *args, **kwargs)\cf2 \
\cf5     782\cf2          set\cf10 .\cf2 \
\cf5     783\cf2          """\
\cf5 --> 784\cf9          \cf5 return\cf2  self\cf10 .\cf2 _filter_or_exclude\cf10 (\cf2 False\cf10 ,\cf2  \cf10 *\cf2 args\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 )\cf2 \
\cf5     785\cf2  \
\cf5     786\cf2      \cf5 def\cf2  exclude\cf10 (\cf2 self\cf10 ,\cf2  \cf10 *\cf2 args\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 ):\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/query.pyc\cf2  in \cf4 _filter_or_exclude\cf10 (self, negate, *args, **kwargs)\cf2 \
\cf5     800\cf2              clone\cf10 .\cf2 query\cf10 .\cf2 add_q\cf10 (~\cf2 Q\cf10 (*\cf2 args\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 ))\cf2 \
\cf5     801\cf2          \cf5 else\cf10 :\cf2 \
\cf5 --> 802\cf9              \cf2 clone\cf10 .\cf2 query\cf10 .\cf2 add_q\cf10 (\cf2 Q\cf10 (*\cf2 args\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 ))\cf2 \
\cf5     803\cf2          \cf5 return\cf2  clone\
\cf5     804\cf2  \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/sql/query.pyc\cf2  in \cf4 add_q\cf10 (self, q_object)\cf2 \
\cf5    1248\cf2          existing_inner = set(\
\cf5    1249\cf2              (a for a in self.alias_map if self.alias_map[a].join_type == INNER))\
\cf5 -> 1250\cf9          \cf2 clause\cf10 ,\cf2  _ \cf10 =\cf2  self\cf10 .\cf2 _add_q\cf10 (\cf2 q_object\cf10 ,\cf2  self\cf10 .\cf2 used_aliases\cf10 )\cf2 \
\cf5    1251\cf2          \cf5 if\cf2  clause\cf10 :\cf2 \
\cf5    1252\cf2              self\cf10 .\cf2 where\cf10 .\cf2 add\cf10 (\cf2 clause\cf10 ,\cf2  AND\cf10 )\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/sql/query.pyc\cf2  in \cf4 _add_q\cf10 (self, q_object, used_aliases, branch_negated, current_negated, allow_joins, split_subq)\cf2 \
\cf5    1274\cf2                      child\cf10 ,\cf2  can_reuse\cf10 =\cf2 used_aliases\cf10 ,\cf2  branch_negated\cf10 =\cf2 branch_negated\cf10 ,\cf2 \
\cf5    1275\cf2                      current_negated\cf10 =\cf2 current_negated\cf10 ,\cf2  connector\cf10 =\cf2 connector\cf10 ,\cf2 \
\cf5 -> 1276\cf9                      \cf2 allow_joins\cf10 =\cf2 allow_joins\cf10 ,\cf2  split_subq\cf10 =\cf2 split_subq\cf10 ,\cf2 \
\cf5    1277\cf2                  )\
\cf5    1278\cf2                  joinpromoter\cf10 .\cf2 add_votes\cf10 (\cf2 needed_inner\cf10 )\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/sql/query.pyc\cf2  in \cf4 build_filter\cf10 (self, filter_expr, branch_negated, current_negated, can_reuse, connector, allow_joins, split_subq)\cf2 \
\cf5    1204\cf2              \cf5 else\cf10 :\cf2 \
\cf5    1205\cf2                  lhs \cf10 =\cf2  MultiColSource\cf10 (\cf2 alias\cf10 ,\cf2  targets\cf10 ,\cf2  sources\cf10 ,\cf2  field\cf10 )\cf2 \
\cf5 -> 1206\cf9              \cf2 condition \cf10 =\cf2  lookup_class\cf10 (\cf2 lhs\cf10 ,\cf2  value\cf10 )\cf2 \
\cf5    1207\cf2              lookup_type \cf10 =\cf2  lookup_class\cf10 .\cf2 lookup_name\
\cf5    1208\cf2          \cf5 else\cf10 :\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/lookups.pyc\cf2  in \cf4 __init__\cf10 (self, lhs, rhs)\cf2 \
\cf5      22\cf2      \cf5 def\cf2  __init__\cf10 (\cf2 self\cf10 ,\cf2  lhs\cf10 ,\cf2  rhs\cf10 ):\cf2 \
\cf5      23\cf2          self\cf10 .\cf2 lhs\cf10 ,\cf2  self\cf10 .\cf2 rhs \cf10 =\cf2  lhs\cf10 ,\cf2  rhs\
\cf5 ---> 24\cf9          \cf2 self\cf10 .\cf2 rhs \cf10 =\cf2  self\cf10 .\cf2 get_prep_lookup\cf10 ()\cf2 \
\cf5      25\cf2          \cf5 if\cf2  hasattr\cf10 (\cf2 self\cf10 .\cf2 lhs\cf10 ,\cf2  \cf10 'get_bilateral_transforms'):\cf2 \
\cf5      26\cf2              bilateral_transforms \cf10 =\cf2  self\cf10 .\cf2 lhs\cf10 .\cf2 get_bilateral_transforms\cf10 ()\cf2 \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/fields/related_lookups.pyc\cf2  in \cf4 get_prep_lookup\cf10 (self)\cf2 \
\cf5      54\cf2                  \cf9 # only one as we don't get to the direct value branch otherwise.\cf2 \
\cf5      55\cf2                  target_field \cf10 =\cf2  self\cf10 .\cf2 lhs\cf10 .\cf2 output_field\cf10 .\cf2 get_path_info\cf10 ()[-\cf4 1\cf10 ].\cf2 target_fields\cf10 [-\cf4 1\cf10 ]\cf2 \
\cf5 ---> 56\cf9                  \cf2 self\cf10 .\cf2 rhs \cf10 =\cf2  \cf10 [\cf2 target_field\cf10 .\cf2 get_prep_value\cf10 (\cf2 v\cf10 )\cf2  \cf5 for\cf2  v \cf5 in\cf2  self\cf10 .\cf2 rhs\cf10 ]\cf2 \
\cf5      57\cf2          \cf5 return\cf2  super\cf10 (\cf2 RelatedIn\cf10 ,\cf2  self\cf10 ).\cf2 get_prep_lookup\cf10 ()\cf2 \
\cf5      58\cf2  \
\
\cf5 /Users/fifihome/Desktop/Python/python_stack/myEnvironments/djangoEnv/lib/python2.7/site-packages/django/db/models/fields/__init__.pyc\cf2  in \cf4 get_prep_value\cf10 (self, value)\cf2 \
\cf5     964\cf2          \cf5 if\cf2  value \cf5 is\cf2  None\cf10 :\cf2 \
\cf5     965\cf2              \cf5 return\cf2  None\
\cf5 --> 966\cf9          \cf5 return\cf2  int\cf10 (\cf2 value\cf10 )\cf2 \
\cf5     967\cf2  \
\cf5     968\cf2      \cf5 def\cf2  contribute_to_class\cf10 (\cf2 self\cf10 ,\cf2  cls\cf10 ,\cf2  name\cf10 ,\cf2  \cf10 **\cf2 kwargs\cf10 ):\cf2 \
\
\cf9 TypeError\cf2 : int() argument must be a string or a number, not 'QuerySet'\
\
\cf6 In [\cf7 53\cf6 ]: \cf2 b4.books.add(a)\
\
\cf6 In [\cf7 54\cf6 ]: \cf2 b4.save()\
\
\cf6 In [\cf7 55\cf6 ]: \cf2 b4.books.add(a1)\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 NameError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-55-07ed599845bf>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 b4\cf10 .\cf2 books\cf10 .\cf2 add\cf10 (\cf2 a1\cf10 )\cf2 \
\
\cf9 NameError\cf2 : name 'a1' is not defined\
\
\cf6 In [\cf7 56\cf6 ]: \cf2 a1 = Book.objects.get(\cf6 id\cf2 =\cf6 2\cf2 )\
\
\cf6 In [\cf7 57\cf6 ]: \cf2 b4.books.add(a2)\
\
\cf6 In [\cf7 58\cf6 ]: \cf2 b4.books.add(a3)\
\
\cf6 In [\cf7 59\cf6 ]: \cf2 b3.books.add(a4)\
\
\cf6 In [\cf7 60\cf6 ]: \cf2 b3.save()\
\
\cf6 In [\cf7 61\cf6 ]: \cf2 Book.objects.get(\cf6 id\cf2 =\cf6 3\cf2 ).books_authors.all()\
\cf12 Out[\cf13 61\cf12 ]: \cf2 <QuerySet [<Author object: Speros >, <Author object: John >, <Author object: Ketul >]>\
\
\cf6 In [\cf7 62\cf6 ]: \cf2 Book.objects.get(\cf6 id\cf2 =\cf6 3\cf2 ).books_authors.delete(b2)\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-62-3903eb886398>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Book\cf10 .\cf2 objects\cf10 .\cf2 get\cf10 (\cf2 id\cf10 =\cf4 3\cf10 ).\cf2 books_authors\cf10 .\cf2 delete\cf10 (\cf2 b2\cf10 )\cf2 \
\
\cf9 AttributeError\cf2 : 'ManyRelatedManager' object has no attribute 'delete'\
\
\cf6 In [\cf7 63\cf6 ]: \cf2 Book.objects.get(\cf6 id\cf2 =\cf6 3\cf2 ).books_authors.delete(\cf6 id\cf2 =\cf6 1\cf2 )\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-63-0ea6968a972b>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Book\cf10 .\cf2 objects\cf10 .\cf2 get\cf10 (\cf2 id\cf10 =\cf4 3\cf10 ).\cf2 books_authors\cf10 .\cf2 delete\cf10 (\cf2 id\cf10 =\cf4 1\cf10 )\cf2 \
\
\cf9 AttributeError\cf2 : 'ManyRelatedManager' object has no attribute 'delete'\
\
\cf6 In [\cf7 64\cf6 ]: \cf2 Book.objects.get(\cf6 id\cf2 =\cf6 3\cf2 )\
\cf12 Out[\cf13 64\cf12 ]: \cf2 <Book object: Python SOME STRING>\
\
\cf6 In [\cf7 65\cf6 ]: \cf2 b2.books.delete(a3)\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-65-a6d4ba2da5b1>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 b2\cf10 .\cf2 books\cf10 .\cf2 delete\cf10 (\cf2 a3\cf10 )\cf2 \
\
\cf9 AttributeError\cf2 : 'ManyRelatedManager' object has no attribute 'delete'\
\
\cf6 In [\cf7 66\cf6 ]: \cf2 b2.books.remove(a3)\
\
\cf6 In [\cf7 67\cf6 ]: \cf2 Book.objects.get(\cf6 id\cf2 =\cf6 3\cf2 ).books_authors.all()\
\cf12 Out[\cf13 67\cf12 ]: \cf2 <QuerySet [<Author object: John >, <Author object: Ketul >]>\
\
\cf6 In [\cf7 68\cf6 ]: \cf2 Author.objects.lst()\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 AttributeError\cf2                             Traceback (most recent call last)\
\cf5 <ipython-input-68-76ea09e951a9>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 Author\cf10 .\cf2 objects\cf10 .\cf2 lst\cf10 ()\cf2 \
\
\cf9 AttributeError\cf2 : 'Manager' object has no attribute 'lst'\
\
\cf6 In [\cf7 69\cf6 ]: \cf2 Author.objects.last()\
\cf12 Out[\cf13 69\cf12 ]: \cf2 <Author object: Ketul >\
\
\cf6 In [\cf7 70\cf6 ]: \cf2 Author.objects.all()\
\cf12 Out[\cf13 70\cf12 ]: \cf2 <QuerySet [<Author object: Mike >, <Author object: Speros >, <Author object: John >, <Author object: Jadee >, <Author object: Ketul >]>\
\
\cf6 In [\cf7 71\cf6 ]: \cf2 b5.books.add(a2)\
\cf9 ---------------------------------------------------------------------------\cf2 \
\cf9 NameError\cf2                                  Traceback (most recent call last)\
\cf5 <ipython-input-71-67f265ad21eb>\cf2  in \cf4 <module>\cf10 ()\cf2 \
\cf5 ----> 1\cf9  \cf2 b5\cf10 .\cf2 books\cf10 .\cf2 add\cf10 (\cf2 a2\cf10 )\cf2 \
\
\cf9 NameError\cf2 : name 'b5' is not defined\
\
\cf6 In [\cf7 72\cf6 ]: \cf2 b4.books.add(a2)\
\
\cf6 In [\cf7 73\cf6 ]: \cf2 b4.save()\
\
\cf6 In [\cf7 74\cf6 ]: \cf2 Book.objects.get(\cf6 id\cf2 =\cf6 2\cf2 ).books_authors.all()\
\cf12 Out[\cf13 74\cf12 ]: \cf2 <QuerySet [<Author object: Mike >, <Author object: Speros >, <Author object: John >, <Author object: Ketul >]>\
\
\cf6 In [\cf7 75\cf6 ]: \cf2 Author.objects.get(\cf6 id\cf2 =\cf6 3\cf2 )books.all()\
\cf4   File \cf5 "<ipython-input-75-ccc1d6fd11b1>"\cf4 , line \cf5 1\cf2 \
\cf9     Author.objects.get(id=3)books.all()\cf2 \
                                ^\
\cf9 SyntaxError:\cf2  invalid syntax\
\
\
\cf6 In [\cf7 76\cf6 ]: \cf2 Author.objects.get(\cf6 id\cf2 =\cf6 3\cf2 ).books.all()\
\cf12 Out[\cf13 76\cf12 ]: \cf2 <QuerySet [<Book object: C Sharp SOME STRING>, <Book object: Java SOME STRING>, <Book object: Python SOME STRING>, <Book object: PHP SOME STRING>]>\
\
\cf6 In [\cf7 77\cf6 ]: \cf2 Author.objects.get(\cf6 id\cf2 =\cf6 2\cf2 ).books.all()\
\cf12 Out[\cf13 77\cf12 ]: \cf2 <QuerySet [<Book object: C Sharp SOME STRING>, <Book object: Java SOME STRING>]>\
\
\cf6 In [\cf7 78\cf6 ]: \cf2  }